### DISCUSSION TITLE
Is there any feature like namespace or scope support plan?

### DISCUSSION THREAD

**Discussion by OXeu** [typst/typst#3179] [General]
### Description

I have a requirement that there are two similar large plates when writing the article. They only have different variable values and a few text differences. I want to reuse this section, but it seems troublesome to encapsulate it (because my text and variables are mixed together, and the text will be calculated with the variables. Part of text would quite different when the calculated value reached some threshold) and it's not necessary (I only have two similar modules), so my idea is to manually modify the variable value and text value after copying and pasting directly. This works very well, but when I write in the follow-up, I will need to use the variables of the first two parts, but due to direct copying and pasting, in the end, only The variables in the second part can be accessed, and all the variables in the first part are overwritten, so is there any feature like namespace or scope? Just like the code below.
```typst
#namespace(section1)[
Some Content
#let a = 1
]
#namespace(section2)[
Some Content
#let a = 2
]
// and we can accessible by the following code
#section1.a // 1
#section2.a // 2
```

### Use Case

Just as the above description

**Comment by OXeu** [comment]
A more natural grammar maybe as follows:
```typst
#let section1 = [
Content...
#let var = 1
Section1 Value: #var
]
#let section2 = [
Content...
#let var = 2
Section2 Value: #var
]

Sum: #(section1.var + section2.var)
```
It only needs to implement variable export.

**Comment by max397574** [comment]
use a function?

**Comment by OXeu** [comment]
I don't know what your exact meaning is, guess you want me to return the needed value by function? That should work for me to solve my current problems. But I still want this feature because I have over 70 variables and I have to pick out what I need (nearly 20 of them). But still thank you for providing a new way I ignored to solve my current problems.







---Original---
From: ***@***.***&gt;
Date: Fri, Jan 12, 2024 01:24 AM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [typst/typst] Is there any feature like namespace or scoopsupport plan? (Issue #3173)




 
use a function?

**Comment by frozolotl** [comment]
You can define multiple separate files and use `import`:

```typ
// section1.typ
#let var = 1

// section2.typ
#let var = 2

// main.typ
#import "section1.typ"
#import "section2.typ"

Sum: #(section1.var + section2.var)
```

Import does not include any content within the files. For that, you need to use `include`. See https://typst.app/docs/reference/scripting/#modules for more information.

**Comment by OXeu** [comment]
> You can define multiple separate files and use `import`:
> 
> ```typst
> // section1.typ
> #let var = 1
> 
> // section2.typ
> #let var = 2
> 
> // main.typ
> #import "section1.typ"
> #import "section2.typ"
> 
> Sum: #(section1.var + section2.var)
> ```
> 
> Import does not include any content within the files. For that, you need to use `include`. See https://typst.app/docs/reference/scripting/#modules for more information.


I tried that, but I found in the section also used many variables defined before. Once I move this section to a separate file I can't use these variables any more.

**Reply by laurmaedje** [reply]
You can put those common things into a third file, which you import in both.

**Reply by OXeu** [reply]
Unfortunately, all calculations are done one after the other. Moreover, the variables are mixed in the text, because I tend to declare and calculate the variables before the text is displayed, and many variables will be changed multiple times in subsequent calculations and verifications (see the example below), making it difficult to pick them out. and put it into a third file, because this will break the context of the variables. To solve this problem, I may need to spend more time using new names for variables that change repeatedly to avoid them being overwritten, and then Then put it into a separate file.

For example, my document might look like this (this is just a very simplified version, the real situation is much more complicated than this):
```typst
#let f = calc.round(calc.cos(1), digits: 3)
Therefore, the minimum force required for work is: #f, which after rounding up is #let f=calc.ceil(f); #f
```
Because my paper involves a huge amount of calculations and intermediate variables, in order to avoid the trouble of naming and subsequent reference to wrong variables, I usually perform new assignment operations directly in the text, although I know this may not be a good idea.
